internal/lang/eval.go

```diff
@@ -224,32 +219,23 @@ func (s *Scope) enhanceFunctionDiags(diags hcl.Diagnostics) hcl.Diagnostics {
 
 			// Update enhanced with additional details
 
-			if fullNamespace == CoreNamespace {
+			fn := addrs.ParseFunction(fullNamespace + funcName)
+
+			if fn.IsNamespace(addrs.FunctionNamespaceCore) {
 				// Error is in core namespace, mirror non-core equivalent
 				enhanced.Summary = "Call to unknown function"
-				enhanced.Detail = fmt.Sprintf("There is no builtin (%s) function named %q.", CoreNamespace, funcName)
-				continue
-			}
-
-			match := providerFuncNamespace.FindSubmatch([]byte(fullNamespace))
-			if match == nil || string(match[1]) != "provider" {
-				// complete mismatch or invalid prefix
-				enhanced.Summary = "Invalid function format"
-				enhanced.Detail = fmt.Sprintf("Expected provider::<provider_name>::<function_name>, instead found \"%s%s\"", fullNamespace, funcName)
-				continue
-			}
-
-			providerName := string(match[2])
-			addr, ok := s.ProviderNames[providerName]
-			if !ok {
-				// Provider not registered
-				enhanced.Summary = "Unknown function provider"
-				enhanced.Detail = fmt.Sprintf("Provider %q does not exist within the required_providers of this module", providerName)
+				enhanced.Detail = fmt.Sprintf("There is no builtin (%s::) function named %q.", addrs.FunctionNamespaceCore, funcName)
+			} else if fn.IsNamespace(addrs.FunctionNamespaceProvider) {
+				if _, err := fn.AsProviderFunction(); err != nil {
+					// complete mismatch or invalid prefix
+					enhanced.Summary = "Invalid function format"
+					enhanced.Detail = err.Error()
+				}
 			} else {
-				// Func not in provider
-				enhanced.Summary = "Function not found in provider"
-				enhanced.Detail = fmt.Sprintf("Function %q was not registered by provider named %q of type %q", funcName, providerName, addr)
+				enhanced.Summary = "Unknown function namespace"
+				enhanced.Detail = fmt.Sprintf("Function %q does not exist within any known namespace", fn)
```

Comment:
Nit: Shouldn't this be more like "The namespace of function %q does not exist" or something like that?

Cause the current frame is "within any known namespace" even though I believe it **contains** the namespace, so it's not like you're iterating over namespaces to check if it exists anywhere.